sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/monitoring.py

```diff
@@ -0,0 +1,70 @@
+"""Lambda function to send a formatted SNS notification when a Batch job fails."""
+
+import os
+
+import boto3
+
+sns_client = boto3.client("sns")
+logs_client = boto3.client("logs")
+
+
+def lambda_handler(event, context):
+    """Lambda handler to send an SNS notification when a Batch job fails.
+
+    Lambda will format the message and retrieve logging from the failed job, before
+    sending a message to the notification service (SNS topic defined by the environment
+    variable "SNS_ARN").
+
+    Parameters
+    ----------
+    event : dict
+        The JSON formatted document with the data required for the
+        lambda function to process. Source event is from AWS Batch.
+    context : obj
+        The context object for the lambda function
+    """
+    # Extract relevant details from the event
+    detail = event.get("detail", {})
+
+    job_name = detail.get("jobName", "Unknown")
+    job_id = detail.get("jobId", "Unknown")
```

Comment:
Shouldn't we still send an email if the event is "batch job failed" even if we don't know the job name or job id? How would you handle the error instead?

Reply:
For sure. I will try my best. I will add some clarification as well.

What I meant above is batch job event will be in this format where it **should** always have 'jobName' or 'jobId' or 'attempts' or 'container' and so on keys. I took out below example batch job event json from `indexer.py` lambda code. It's missing some other fields because I only kept information that indexer cares about. But batch event job will always has fixed keys.
```
{
        "detail-type": "Batch Job State Change",
        "source": "aws.batch",
        "detail": {
            "jobArn": (
                "arn:aws:batch:us-west-2:012345678910:"
                "job/26242c7e-3d49-4e41-9387-74fcaf9630bb"
            ),
            "jobName": "swe-l0-job",
            "jobId": "26242c7e-3d49-4e41-9387-74fcaf9630bb",
            "jobQueue": (
                "arn:aws:batch:us-west-2:012345678910:"
                "job-queue/swe-fargate-batch-job-queue"
            ),
            "status": "FAILED",
            "statusReason": "some error message",
            "jobDefinition": (
                "arn:aws:batch:us-west-2:012345678910:"
                "job-definition/fargate-batch-job-definitionswe:1"
            ),
            "container": {
                "image": (
                    "123456789012.dkr.ecr.us-west-2.amazonaws.com/" "swapi-repo:latest"
                ),
                "command": [
                    "--instrument", "swapi",
                    "--data-level", "l1",
                    "--descriptor", "sci",
                    "--start-date", "20230724",
                    "--version", "v001",
                    "--dependency", \"""[
                        {
                            'instrument': 'swapi',
                            'level': 'l0',
                            'start_date': 20230724,
                            'version': 'v001'
                        }
                    ]\""",
                    "--upload-to-sdc",
                ],
                "logStreamName": (
                    "fargate-batch-job-definitionswe/default/"
                    "8a2b784c7bd342f69ea5dac3adaed26f"
                ),
            },
        }
    }
```

Given above reason, we should never receive batch job event json that didn't have `jobName` or `jobId` and so on. If we do, then it's a bug and we need to correct our event source or lambda trigger source and update accordingly. I hope that helps add some context I missed to add last time.

Reply:
To your question about how we should handle error, I would check event source. If it's not 'aws.batch', then raise error and shouldn't send message to the SNS. That's my thought.

Reply:
Ok, that makes sense! I'll add that case.